java-checks/src/main/java/org/sonar/java/checks/unused/UnusedPrivateMethodCheck.java

```diff
@@ -206,4 +220,41 @@ private static boolean isConstructorWithParameters(MethodTree methodTree) {
   private static boolean isNotMethodFromSerializable(MethodTree methodTree, Symbol symbol) {
     return methodTree.is(Tree.Kind.METHOD) && !SerializableContract.SERIALIZABLE_CONTRACT_METHODS.contains(symbol.name());
   }
+
+  private static class CheckAnnotationsForMethodNames extends BaseTreeVisitor {
+
+    public CheckAnnotationsForMethodNames(Set<String> methodNames) {
+      this.methodNames = methodNames;
+    }
+
+    private Set<String> methodNames;
```

Comment:
Using a `Set` here is not sufficient: if in the same file you have two different classes like so
```
  @MethodProvided(value = "foo")
  static class TN {
    private void foo() {} // Compliant, method is referenced in annotation
  }
  
  static class FN {
    private void foo() {} // False negative, this method has same name but no references
  }
```

the `CheckAnnotationsForMethodNames` will remove the name `foo` once and for all and the second encounter will not trigger an issue.

Maybe you could use the whole method signature instead of the simple name? Or a different data structure?